src/Orleans/CodeGeneration/CodeGeneratorManager.cs

```diff
@@ -30,20 +32,27 @@ internal static class CodeGeneratorManager
         /// <summary>
         /// The runtime code generator.
         /// </summary>
-        private static IRuntimeCodeGenerator codeGeneratorInstance;
+        private IRuntimeCodeGenerator codeGeneratorInstance;
 
         /// <summary>
         /// The code generator cache.
         /// </summary>
         private static ICodeGeneratorCache codeGeneratorCacheInstance;
 
+        private readonly IServiceProvider serviceProvider;
+
+        public CodeGeneratorManager(IServiceProvider serviceProvider)
+        {
+            this.serviceProvider = serviceProvider;
+        }
+
         /// <summary>
         /// Loads the code generator on demand
         /// </summary>
-        public static void Initialize()
+        public void Initialize()
         {
-            codeGeneratorInstance = LoadCodeGenerator();
-            codeGeneratorCacheInstance = codeGeneratorInstance as ICodeGeneratorCache;
+            this.codeGeneratorInstance = this.serviceProvider.GetService<IRuntimeCodeGenerator>() ?? this.LoadCodeGenerator();
```

Comment:
If the user injected a code generator, that should be used instead of loading the default assembly.